### PR TITLE
Allow templated functors in parallel_for, parallel_reduce and parallel_scan

### PR DESCRIPTION
**Pull request by masterleinad**
Part of #5908. Fixes https://github.com/kokkos/kokkos/issues/5156, improves https://github.com/kokkos/kokkos/issues/2034. This is basically #5420:

Extending `FunctorAnalysis` to be able to be used with templated call operators appears to be very difficult. On the other hand, the overloads of `parallel_reduce` and `parallel_scan` taking a return value already contain a hint for the `value_type` to use so that we can avoid using the deduction in `FunctorAnalysis` in those cases.

This pull request does exactly that. If the functor has a `value_type` alias, we use that, otherwise, we use the `value_type` deduced from the return type and if that is void, too, we do the deduction in FunctorAnalysis based on the functor's call operator.

edit: Copied description from #5420.

### CODE REVIEW

File: core/src/impl/Kokkos_FunctorAnalysis.hpp [hunk 1]
```diff
@@ -126,7 +127,7 @@ struct FunctorAnalysis {
 
   template <typename F, typename = std::false_type>
   struct has_value_type {
-    using type = void;
+    using type = OverrideValueType;
```
Comment (masterleinad) [inline core/src/impl/Kokkos_FunctorAnalysis.hpp:130]: This is the most important change. Passing in a non-void `OverrideValueType` skips the value type deduction.

**Comment by dalg24** [comment]
Please resolve the conflicts and improve the description

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 2]
```diff
@@ -477,22 +477,24 @@ __device__ bool hip_single_inter_block_reduce_scan(
 }
 
 // Size in bytes required for inter block reduce or scan
-template <bool DoScan, class FunctorType, class ArgTag>
+template <bool DoScan, class ArgTag, class ValueType, class FunctorType>
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp:480]: Why did you change the other of the template parameters?  You didn't change it in the CUDA backend.

Comment (masterleinad) [reply core/src/HIP/Kokkos_HIP_ReduceScan.hpp:480]: So I can take advantage of template argument deduction. I can mirror for Cuda if you prefer.

Comment (Rombur) [reply core/src/HIP/Kokkos_HIP_ReduceScan.hpp:480]: Yes, please.

File: core/unit_test/TestFunctorAnalysis.hpp [hunk 3]
```diff
@@ -85,7 +85,7 @@ void test_functor_analysis() {
   TestFunctorAnalysis_03 c03;
   using A03 = Kokkos::Impl::FunctorAnalysis<
       Kokkos::Impl::FunctorPatternInterface::REDUCE,
-      Kokkos::RangePolicy<ExecSpace>, TestFunctorAnalysis_03>;
+      Kokkos::RangePolicy<ExecSpace>, TestFunctorAnalysis_03, void>;
```
Comment (tcclevenger) [inline core/unit_test/TestFunctorAnalysis.hpp:88]: Should a non-void `OverrideValueType` test be added where we static_assert the value type is the override?

**Comment by Rombur** [comment]
clang-format is not happy

File: core/unit_test/TestFunctorAnalysis.hpp (hunk 3 above)
Comment (masterleinad) [reply core/unit_test/TestFunctorAnalysis.hpp:98]: Sure. I added another test.

**Comment by masterleinad** [comment]
> clang-format is not happy

Fixed.

**Review by Rombur — APPROVED** [review]

**Review by tcclevenger — APPROVED** [review]

**Review by nliber — APPROVED** [review]

**Comment by masterleinad** [comment]
Only `HIP-ROCm-5.2-C++20` is failing (it seems the machine was broken).

**Comment by dalg24** [comment]
I don't like the complexity with the "lambda factory" capturing everything by reference but others seems to think it is fine.

**Review by dalg24 — COMMENTED** [review]
Just posting pending comments that I had.  Not a full review.

File: core/src/impl/Kokkos_FunctorAnalysis.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_FunctorAnalysis.hpp:131]: Done.

File: core/unit_test/incremental/Test05_ParallelReduce_RangePolicy.hpp [hunk 4]
```diff
@@ -28,8 +28,9 @@ using value_type       = double;
 constexpr double value = 0.5;
 
 struct ReduceFunctor {
-  KOKKOS_INLINE_FUNCTION
-  void operator()(const int i, double &UpdateSum) const {
+  template <typename SizeType, typename ValueType>
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/incremental/Test05_ParallelReduce_RangePolicy.hpp]: Done.

File: core/unit_test/incremental/Test16_ParallelScan.hpp [hunk 5]
```diff
@@ -61,19 +61,34 @@ struct NonTrivialScanFunctor {
   ~NonTrivialScanFunctor() {}
 };
 
+template <typename ExecSpace>
+struct GenericScanFunctor {
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/incremental/Test16_ParallelScan.hpp]: Done.

File: core/unit_test/incremental/Test16_ParallelScan.hpp [hunk 6]
```diff
@@ -61,19 +61,34 @@ struct NonTrivialScanFunctor {
...
+  Kokkos::View<value_type *, ExecSpace> d_data;
+
+  template <typename SizeType, typename ValueType>
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/incremental/Test16_ParallelScan.hpp]: Done.

File: core/unit_test/incremental/Test16_ParallelScan.hpp [hunk 7]
```diff
@@ -61,19 +61,34 @@ struct NonTrivialScanFunctor {
...
+    update_value += val_i;
+  }
+};
+
 template <class ExecSpace>
 struct TestScan {
   // 1D  View of double
-  using View_1D = typename Kokkos::View<value_type *, ExecSpace>;
+  using View_1D  = typename Kokkos::View<value_type *, ExecSpace>;
+  View_1D d_data = View_1D("data", N);
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/incremental/Test16_ParallelScan.hpp]: Let me revert this.

File: core/unit_test/incremental/Test16_ParallelScan.hpp [hunk 8]
```diff
@@ -77,18 +77,16 @@ struct GenericScanFunctor {
...
 struct TestScan {
   // 1D  View of double
-  using View_1D  = typename Kokkos::View<value_type *, ExecSpace>;
-  View_1D d_data = View_1D("data", N);
-
-  template <typename SizeType>
-  KOKKOS_FUNCTION void operator()(SizeType i) const {
-    d_data(i) = i * 0.5;
-  }
+  using View_1D = typename Kokkos::View<value_type *, ExecSpace>;
```
Comment (dalg24) [inline core/unit_test/incremental/Test16_ParallelScan.hpp]: `typename` is superfluous here

Comment (dalg24) [reply core/unit_test/incremental/Test16_ParallelScan.hpp]: Fixing is optional. Will merge if the CI passes.
